### ISSUE TITLE
Black Box Autosave

### ISSUE THREAD

**Issue by J-Shep** [CorsixTH/CorsixTH#836]
Hi,

With errors which can't be reproduced being very difficult to fix I think having a black box autosave overwritten every X seconds, when no error have occurred, would be very useful because if it was overwritten every 5 seconds or less then it would probably be able to reproduce an error without the tester being required to do anything. 

A player, developer or beta tester could change the frequency of this autosave with a setting in their settings menu and they could have the option to disable it as well. 

The default settings for this black box autosave could depend on if they are for a trunk, beta or release version. So for trunk and beta versions the default save frequency could be 1 second where as it could be 5 seconds for released versions.

I don't think this saved game should be listed in the saved game drop down menu because with it being so frequent a player wouldn't be able to undo any of their decisions with it.

I will quickly implement this idea if it's accepted.

**Comment by bobcanthelpyou** [comment]
I also thought about something like this, cause it is always hard to have a savegame before a problem occurs, sometimes not critical error are seen to late and can't be reproduced or no autosave is available anymore. 

I just thought about something like using the normal autosave but increase the autosave number until the end of the game. Also there should be different autosaves for different game (autosaves should not be overwritten).

But the save interval should not be less than a minute, when the game is saving all the time you can't play it. And you will very fast run into problems with free space on your hard drive

Save interval 1 second, for one hour, autosave size 0,5 MB:
60 * 60 * 0,5 = 1800 MB

Also this should not be enabled by default.

**Comment by J-Shep** [comment]
I opened a issue with the same autosave improvement ideas that you've suggested in the past but its closed now, I think I may have closed it. But I now think this issue should be reopened because the ideas you've suggested would be worthwhile improvement imo.

However one minute auto saves would be very unlikely to be able to reproduce errors without testers doing anything and auto saves may already be being made in less than one minute intervals because they happen every X game months, so the auto save rate is related to the game speed.

I personally think most people would have computers with modern enough hard drives & SSDs to easily handle autosaving CorsixTH every 5 seconds because this simple 2D game doesn't have much data to write to its saved game files compared to more complicated 3D games. And players who don't have modern enough hardware could change their blackbox auto save rate in their settings or could just disable this autosave.

I will test how different computers I own cope with autosaving CorsixTH every 5 seconds.

**Comment by bobcanthelpyou** [comment]
> For a single "black box" autosave being overwritten every X seconds when its updated, this wouldn't be a problem. 

This is only a single savegame? And this is only for detecting crash errors? Not for other problems that don't crash the game?

**Comment by Alberth289346** [comment]
I wonder if you could save to memory instead of disk. In my experience, while Lua may crash, the program itself usually keeps on running. When a crash is detected, you can save whatever was saved last.
Btw maybe you want 2 copies, and alternately save to the other copy. That way you still have a copy if the save itself crashes.

**Comment by J-Shep** [comment]
>> This is only a single savegame? And this is only for detecting crash errors? Not for other problems that don't crash the game?

@bobcanthelpyou Yes, my idea is for there to be a single "black_box.sav" autosave feature which would have a different purpose than the existing autosave feature: to help developers reproduce crash errors without having to know what input events caused them in order to reproduce them. 

So the save rate for this new & seperate autosave feature shouldn't be related to the game speed because its purpose wouldn't be to allow players to restore a hospital to the state it was in at a previous point in its world history for the purpose of undoing gameplay decisions.

So there should a separate Github issue for discussing improvements to the existing autosave feature while #836 is for discussing how the new "black box" autosave system should operate if it's implemented.

>> I wonder if you could save to memory instead of disk. In my experience, while Lua may crash, the program itself usually keeps on running. When a crash is detected, you can save to disk whatever was saved last.

@Alberth289346 Thanks for posting this good suggestion: saving black box data to RAM instead of the disk would be ideal because the biggest and busiest of hospitals CorsixTH allows players to make requires few MB of data and buying GB of RAM is cheap nowadays.  I bet even players with old computers & sold hard drives would have no problem having a 5 seconds or quicker black box save rate if its data was stored in RAM and then only written to a hard drive/SSD when the game has crashed, so that crash victim can post this saved game in a Github issue for their problem or the WIP auto error reporting system can do this for them along with all the other saved games their hospital would have from before the error..

>> Btw maybe you want 2 copies, and alternately save to the other copy. That way you still have a copy if the save itself crashes.

Yes your right, thanks for pointing this out.

**Comment by J-Shep** [comment]
>> I wonder if you could save to memory instead of disk. In my experience, while Lua may crash, the program itself usually keeps on running. When a crash is detected, you can save to disk whatever was saved last.

I think the default save location should be in RAM and then in the settings players/beta testers could choose to allow CorsixTH to save black box data to their fast hard drive/SSD instead to ensure that CorsxTH can save its state from shortly before any crash.

**Comment by J-Shep** [comment]
@bobcanthelpyou I've reopened the issue #662 I closed about improving how games are saved automatically with the existing autosave feature and how they are saved manually.

**Comment by J-Shep** [comment]
I think a much more hardware efficient way to ensure that Lua errors can be reproduced would be to just log game state & view changing input events in a replayable way. So black box replays could be recorded between autosaves made by the existing autosave feature in case a Lua error occurs.

I had this idea in the past and opened an issue for it but then I decided this feature wasn't worth implementing so I stopped working on it and closed this Github issue I had opened. But I now think this feature would be worth implementing to help reproduce Lua errors. #589
